File: app/src/main/java/me/echeung/moemoekyun/ui/view/SongList.java
package me.echeung.moemoekyun.ui.view;

import android.app.Activity;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.PopupMenu;
import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.MenuItem;
import android.widget.Toast;

import java.lang.ref.WeakReference;

import me.echeung.moemoekyun.R;
import me.echeung.moemoekyun.adapter.SongAdapter;
import me.echeung.moemoekyun.client.model.Song;
import me.echeung.moemoekyun.databinding.SongsListBinding;
import me.echeung.moemoekyun.util.SongActionsUtil;
import me.echeung.moemoekyun.util.SongSortUtil;
import me.echeung.moemoekyun.viewmodel.SongListViewModel;

public class SongList {

    private WeakReference<Activity> activity;
    private SongAdapter adapter;
    private SongListLoader loader;
    private SwipeRefreshLayout swipeRefreshLayout;

    private SongListViewModel songListViewModel;

    public SongList(Activity activity, SongsListBinding binding, String listId, SongListLoader loader) {
        this.activity = new WeakReference<>(activity);
        this.loader = loader;

        // List adapter
        this.adapter = new SongAdapter(activity, listId);
        RecyclerView songsList = binding.list;
        songsList.setLayoutManager(new LinearLayoutManager(activity));
        songsList.setAdapter(adapter);

        // Pull to refresh
        swipeRefreshLayout = binding.refreshLayout;
        swipeRefreshLayout.setColorSchemeResources(R.color.colorAccent);
        swipeRefreshLayout.setOnRefreshListener(this::loadSongs);
        swipeRefreshLayout.setRefreshing(false);

        // Only allow pull to refresh when user is at the top of the list
        songsList.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
            }

            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                int topRowVerticalPosition = songsList.getChildCount() != 0
                        ? songsList.getChildAt(0).getTop()
                        : 0;
                swipeRefreshLayout.setEnabled(topRowVerticalPosition >= 0);
            }
        });

        this.songListViewModel = new SongListViewModel();
        binding.setVm(songListViewModel);

        // Filter
        binding.query.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void afterTextChanged(Editable editable) {
                String query = editable.toString().trim().toLowerCase();
                adapter.filter(query);

                boolean hasResults = adapter.getItemCount() != 0;
                songListViewModel.setHasResults(hasResults);
                if (hasResults) {
                    binding.list.scrollToPosition(0);
                }
            }
        });

        // Menu
        binding.overflowBtn.setOnClickListener(v -> {
            Activity activityRef = this.activity.get();
            if (activityRef == null) return;

            PopupMenu popupMenu = new PopupMenu(activityRef, binding.overflowBtn);
            popupMenu.inflate(R.menu.menu_sort);

            SongSortUtil.initSortMenu(activityRef, listId, popupMenu.getMenu());
            popupMenu.setOnMenuItemClickListener(this::handleMenuItemClick);
            popupMenu.show();
        });
    }

    public void loadSongs() {
        loader.loadSongs(adapter);
    }

    public void showLoading(boolean loading) {
        if (swipeRefreshLayout != null) {
            swipeRefreshLayout.setRefreshing(loading);
        }
    }

    public void notifyDataSetChanged() {
        adapter.notifyDataSetChanged();
    }

    private boolean handleMenuItemClick(MenuItem item) {
        Activity activityRef = activity.get();
        if (activityRef == null) return false;

        if (SongSortUtil.handleSortMenuItem(item, adapter)) {
            return true;
        }

        if (item.getItemId() == R.id.action_random_request) {
            Song randomSong = adapter.getRandomRequestSong();
            if (randomSong != null) {
                SongActionsUtil.request(activityRef, randomSong);
            } else {
                Toast.makeText(activityRef, activityRef.getString(R.string.all_cooldown), Toast.LENGTH_SHORT).show();
            }
            return true;
        }

        return false;
    }

    public interface SongListLoader {
        void loadSongs(SongAdapter adapter);
    }

}

File: app/src/main/kotlin/me/echeung/moemoekyun/ui/view/SongList.kt
package me.echeung.moemoekyun.ui.view

import android.app.Activity
import android.text.Editable
import android.text.TextWatcher
import android.view.MenuItem
import android.view.View
import android.widget.EditText
import androidx.appcompat.widget.SearchView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import me.echeung.moemoekyun.R
import me.echeung.moemoekyun.adapter.SongsListAdapter
import me.echeung.moemoekyun.util.SongActionsUtil
import me.echeung.moemoekyun.util.SongSortUtil
import me.echeung.moemoekyun.viewmodel.SongListViewModel
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import java.lang.ref.WeakReference

class SongList(
    activity: Activity,
    private val songListViewModel: SongListViewModel,
    private val songsList: RecyclerView,
    private val swipeRefreshLayout: SwipeRefreshLayout?,
    filterView: View,
    listId: String,
    private val loadSongs: (adapter: SongsListAdapter) -> Unit
) : KoinComponent {

    private val songActionsUtil: SongActionsUtil by inject()
    private val songSortUtil: SongSortUtil by inject()

    private val activity: WeakReference<Activity> = WeakReference(activity)
    private val adapter: SongsListAdapter = SongsListAdapter(activity, listId)

    init {
        // List adapter
        songsList.layoutManager = LinearLayoutManager(activity)
        songsList.adapter = adapter

        // Pull to refresh
        if (swipeRefreshLayout != null) {
            swipeRefreshLayout.setColorSchemeResources(R.color.colorAccent)
            swipeRefreshLayout.setOnRefreshListener(SwipeRefreshLayout.OnRefreshListener { this.loadSongs() })
            swipeRefreshLayout.isRefreshing = false

            // Only allow pull to refresh when user is at the top of the list
            songsList.addOnScrollListener(object : RecyclerView.OnScrollListener() {
                override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                    val topRowVerticalPosition = if (songsList.childCount != 0) {
                        songsList.getChildAt(0).top
                    } else {
                        0
                    }
                    swipeRefreshLayout.isEnabled = topRowVerticalPosition >= 0
                }
            })
        }

        // Filter
        if (filterView is EditText) {
            filterView.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}

                override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}

                override fun afterTextChanged(editable: Editable) {
                    handleQuery(editable.toString())
                }
            })
        }

        if (filterView is SearchView) {
            filterView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
                override fun onQueryTextSubmit(query: String?): Boolean {
                    handleQuery(query!!)
                    return true
                }

                override fun onQueryTextChange(newText: String?): Boolean {
                    handleQuery(newText!!)
                    return true
                }
            })

            filterView.setOnCloseListener {
                handleQuery("")
                true
            }
        }
    }

    private fun handleQuery(query: String) {
        val trimmedQuery = query.trim { it <= ' ' }.lowercase()

        adapter.filter(trimmedQuery)

        val hasResults = adapter.itemCount != 0
        songListViewModel.hasResults = hasResults
        if (hasResults) {
            songsList.scrollToPosition(0)
        }
    }

    fun loadSongs() {
        loadSongs(adapter)
    }

    fun showLoading(loading: Boolean) {
        swipeRefreshLayout?.isRefreshing = loading
    }

    fun notifyDataSetChanged() {
        adapter.notifyDataSetChanged()
    }

    fun handleMenuItemClick(item: MenuItem): Boolean {
        val activityRef = activity.get() ?: return false

        if (songSortUtil.handleSortMenuItem(item, adapter)) {
            return true
        }

        if (item.itemId == R.id.action_random_request) {
            adapter.randomRequestSong?.let {
                songActionsUtil.request(activityRef, it)
            }
            return true
        }

        return false
    }
}
